### ISSUE TITLE
Crash on startup on Android 13

### ISSUE THREAD

**Issue by wingio** [MuntashirAkon/AppManager#738]
- [x] I know what my device, OS and App Manager versions are
- [x] I know how to take logs
- [x] I know how to reproduce the issue which may not be specific to my device

**Describe the bug**
App instantly crashes on startup.

**To Reproduce**
Steps to reproduce the behaviour:
1. Launch the app

**Expected behavior**
Starts and shows the list of apps

**Crash logs**
```04-26 22:04:56.001 15664 15703 E AndroidRuntime: FATAL EXCEPTION: Thread-3
04-26 22:04:56.001 15664 15703 E AndroidRuntime: Process: io.github.muntashirakon.AppManager, PID: 15664
04-26 22:04:56.001 15664 15703 E AndroidRuntime: java.lang.NoSuchMethodError: No interface method getInstalledPackages(II)Landroid/content/pm/ParceledListSlice; in class Landroid/content/pm/IPackageManager; or its super classes (declaration of 'android.content.pm.IPackageManager' appears in /system/framework/framework.jar)
04-26 22:04:56.001 15664 15703 E AndroidRuntime:        at io.github.muntashirakon.AppManager.servermanager.PackageManagerCompat.getInstalledPackages(PackageManagerCompat.java:90)
04-26 22:04:56.001 15664 15703 E AndroidRuntime:        at io.github.muntashirakon.AppManager.utils.PackageUtils.updateInstalledOrBackedUpApplications(PackageUtils.java:252)
04-26 22:04:56.001 15664 15703 E AndroidRuntime:        at io.github.muntashirakon.AppManager.utils.PackageUtils.lambda$getInstalledOrBackedUpApplicationsFromDb$0(PackageUtils.java:164)
04-26 22:04:56.001 15664 15703 E AndroidRuntime:        at io.github.muntashirakon.AppManager.utils.-$$Lambda$PackageUtils$FDXMpcfHZMTjLENDE7nHv1buIR0.run(Unknown Source:4)
04-26 22:04:56.001 15664 15703 E AndroidRuntime:        at java.lang.Thread.run(Thread.java:1012)```

**Device info**
 - Device: Pixel 5
 - OS Version: Android 13 (Public Beta 1)
 - App Manager Version: 2.6.5.1
 - Mode: no-root

**Comment by MuntashirAkon** [comment]
Can you upload **framework.jar**?

**Comment by wingio** [comment]
https://cdn.discordapp.com/attachments/912089000238514196/969292758101659708/framework.jar here you go

**Comment by MuntashirAkon** [comment]
Should be fixed in b68da215abc5353fad7cb9de625734daeb52f041

**Comment by nate-moo** [comment]
@MuntashirAkon just tested the latest debug build and no installed apps are showing up, it's just continuously loading. I can send logs and stuff if you need.

Android version: 13 beta 1 (TPB1.220310.029)
Phone: Pixel 6

**Comment by MuntashirAkon** [comment]
Logs would definitely help.

**Comment by nate-moo** [comment]
https://storage.googleapis.com/cruddyxyz-bucket/c961ba99-1b3b-4016-9ec9-262a2cd5cdd1.txt
Logs ^^

If there is anything else you would like me to do and/or check I can

**Comment by MuntashirAkon** [comment]
> https://storage.googleapis.com/cruddyxyz-bucket/c961ba99-1b3b-4016-9ec9-262a2cd5cdd1.txt

These logs aren't complete because the logs were filtered using App Manager's package name. Package name-based filtering never works. You have to do a process-based filtering instead. Open running apps page in App Manager, find AM Debug, click on the three-dots menu on the right hand side and click view logs.

**Comment by nate-moo** [comment]
Since I can't load any apps in the main view or any others, would you like me to just send unfiltered via dm or something else?

I occasionally get a similar crash message after the ap has been open for a while:

```
java.lang.NoSuchMethodError: No interface method getPackageInfo(Ljava/lang/String;II)Landroid/content/pm/PackageInfo; in class Landroid/content/pm/IPackageManager; or its super classes (declaration of 'android.content.pm.IPackageManager' appears in /system/framework/framework.jar)
    at io.github.muntashirakon.AppManager.servermanager.PackageManagerCompat.getPackageInfoInternal(PackageManagerCompat.java:354)
    at io.github.muntashirakon.AppManager.servermanager.PackageManagerCompat.getPackageInfo(PackageManagerCompat.java:123)
    at io.github.muntashirakon.AppManager.main.MainViewModel.getNewApplicationItem(MainViewModel.java:583)
    at io.github.muntashirakon.AppManager.main.MainViewModel.updateInfoForPackages(MainViewModel.java:500)
    at io.github.muntashirakon.AppManager.main.MainViewModel.updateInfoForUid(MainViewModel.java:460)
    at io.github.muntashirakon.AppManager.main.MainViewModel.access$000(MainViewModel.java:62)
    at io.github.muntashirakon.AppManager.main.MainViewModel$PackageIntentReceiver.onPackageChanged(MainViewModel.java:742)
    at io.github.muntashirakon.AppManager.types.PackageChangeReceiver$ReceiverHandler.handleMessage(PackageChangeReceiver.java:94)
    at android.os.Handler.dispatchMessage(Handler.java:106)
    at android.os.Looper.loopOnce(Looper.java:201)
    at android.os.Looper.loop(Looper.java:288)
    at android.os.HandlerThread.run(HandlerThread.java:67)

Device Info:
App version: 3.0.0-alpha02-DEBUG
App version code: 401
Android build version: 8473704
Android release version: 12
Android SDK version: 32
Android build ID: TPB1.220310.029
Device brand: google
Device manufacturer: Google
Device name: oriole
Device model: Pixel 6
Device product name: oriole
Device hardware name: oriole
ABIs: [arm64-v8a, armeabi-v7a, armeabi]
ABIs (32bit): [armeabi-v7a, armeabi]
ABIs (64bit): [arm64-v8a]
System language: en-US
In-App Language: auto
Mode: auto
```

**Comment by MuntashirAkon** [comment]
> Android release version: 12
> Android SDK version: 32

This is where the problem is, in this case. SDK version for public beta is 100 (> 32) but it's not the case here.

**Comment by MuntashirAkon** [comment]
> Since I can't load any apps in the main view or any others, would you like me to just send unfiltered via dm or something else?

OK. You can send logs via Email.

**Comment by nate-moo** [comment]
Sent logs, is the version reporting being 12 normal for a beta to try and keep compatibility with apps who check? I checked another app and its also reporting 12 as the version. Settings reports the version as 13 though.

**Comment by nate-moo** [comment]
Ok, I found this: it seems like that API is being depricated in 33 but what's confusing is it seems like it's the same method, just it uses package manager flags instead of an int

https://developer.android.com/reference/android/content/pm/PackageManager.html#getInstalledPackages(int)

**Comment by MuntashirAkon** [comment]
> is the version reporting being 12 normal for a beta to try and keep compatibility with apps who check?

I've never installed beta version of Android, so I don't know. But this is completely unnecessary since SDK for A13 (Tiramisu) has already been released, and it's set to 100 for now, which is way higher than 33. Could also be a mistake: I've read in the news that the latest A12 patch users are unable to switch to A13 because A13 was released before that patch. But I do not know for sure.

> it seems like that API is being depricated in 33 but what's confusing is it seems like it's the same method, just it uses package manager flags instead of an int

It's deprecated in `PackageManager` to keep compatibility, but it's actually removed from the internal API (`IPackageManager` or `PackageManagerService`) which is used by App Manager, and compatibility from Android Lollipop to the latest is handled inside App Manager instead.

This can still be fixed though: At first, I have to execute the old method and if I get a `NoSuchMethodError` in A12, I need to try the fallback method (for A12) instead. But exceptions can be a bit costly in this case since the methods are executed many times.

**Comment by nate-moo** [comment]
Instead of checking every time would you be able to check the existence of the method on the start of the app to determine the method to call for the package operations? 

Also, Thanks for taking the time to answer with such a detailed response it helped a lot with understanding the current situation.

**Comment by MuntashirAkon** [comment]
Might be fixed in d13955f389b810a24f60bffc3bf56ed7028bed54

**Comment by MuntashirAkon** [comment]
Did it work?

**Comment by nate-moo** [comment]
Yup! Sorry about the wait, I got pretty busy. Thanks so much for your work!

**Comment by nate-moo** [comment]
Ok nvm it crashes when I try to look at the app permissions and ops
<details>
<summary>log</summary>
`
java.lang.NoClassDefFoundError: io.github.muntashirakon.AppManager.appops.AppOpsManager
    at io.github.muntashirakon.AppManager.details.AppDetailsFragment.permAppOp(AppDetailsFragment.java:625)
    at io.github.muntashirakon.AppManager.details.AppDetailsFragment.access$1000(AppDetailsFragment.java:113)
    at io.github.muntashirakon.AppManager.details.AppDetailsFragment$AppDetailsRecyclerAdapter.getPermissionsView(AppDetailsFragment.java:1556)
    at io.github.muntashirakon.AppManager.details.AppDetailsFragment$AppDetailsRecyclerAdapter.onBindViewHolder(AppDetailsFragment.java:866)
    at io.github.muntashirakon.AppManager.details.AppDetailsFragment$AppDetailsRecyclerAdapter.onBindViewHolder(AppDetailsFragment.java:629)
    at androidx.recyclerview.widget.RecyclerView$Adapter.onBindViewHolder(RecyclerView.java:7254)
    at androidx.recyclerview.widget.RecyclerView$Adapter.bindViewHolder(RecyclerView.java:7337)
    at androidx.recyclerview.widget.RecyclerView$Recycler.tryBindViewHolderByDeadline(RecyclerView.java:6194)
    at androidx.recyclerview.widget.RecyclerView$Recycler.tryGetViewHolderForPositionByDeadline(RecyclerView.java:6460)
    at androidx.recyclerview.widget.RecyclerView$Recycler.getViewForPosition(RecyclerView.java:6300)
    at androidx.recyclerview.widget.RecyclerView$Recycler.getViewForPosition(RecyclerView.java:6296)
    at androidx.recyclerview.widget.LinearLayoutManager$LayoutState.next(LinearLayoutManager.java:2330)
    at androidx.recyclerview.widget.LinearLayoutManager.layoutChunk(LinearLayoutManager.java:1631)
    at androidx.recyclerview.widget.LinearLayoutManager.fill(LinearLayoutManager.java:1591)
    at androidx.recyclerview.widget.LinearLayoutManager.onLayoutChildren(LinearLayoutManager.java:668)
    at androidx.recyclerview.widget.RecyclerView.dispatchLayoutStep2(RecyclerView.java:4309)
    at androidx.recyclerview.widget.RecyclerView.dispatchLayout(RecyclerView.java:4012)
    at androidx.recyclerview.widget.RecyclerView.onLayout(RecyclerView.java:4578)
    at android.view.View.layout(View.java:23579)
    at android.view.ViewGroup.layout(ViewGroup.java:6412)
    at android.widget.FrameLayout.layoutChildren(FrameLayout.java:332)
    at android.widget.FrameLayout.onLayout(FrameLayout.java:270)
    at android.view.View.layout(View.java:23579)
    at android.view.ViewGroup.layout(ViewGroup.java:6412)
    at androidx.appcompat.widget.LinearLayoutCompat.setChildFrame(LinearLayoutCompat.java:1669)
    at androidx.appcompat.widget.LinearLayoutCompat.layoutVertical(LinearLayoutCompat.java:1523)
    at androidx.appcompat.widget.LinearLayoutCompat.onLayout(LinearLayoutCompat.java:1431)
    at android.view.View.layout(View.java:23579)
    at android.view.ViewGroup.layout(ViewGroup.java:6412)
    at androidx.swiperefreshlayout.widget.SwipeRefreshLayout.onLayout(SwipeRefreshLayout.java:688)
    at android.view.View.layout(View.java:23579)
    at android.view.ViewGroup.layout(ViewGroup.java:6412)
    at androidx.viewpager.widget.ViewPager.onLayout(ViewPager.java:1775)
    at android.view.View.layout(View.java:23579)
    at android.view.ViewGroup.layout(ViewGroup.java:6412)
    at com.google.android.material.appbar.HeaderScrollingViewBehavior.layoutChild(HeaderScrollingViewBehavior.java:148)
    at com.google.android.material.appbar.ViewOffsetBehavior.onLayoutChild(ViewOffsetBehavior.java:43)
    at com.google.android.material.appbar.AppBarLayout$ScrollingViewBehavior.onLayoutChild(AppBarLayout.java:2172)
    at androidx.coordinatorlayout.widget.CoordinatorLayout.onLayout(CoordinatorLayout.java:918)
    at android.view.View.layout(View.java:23579)
    at android.view.ViewGroup.layout(ViewGroup.java:6412)
    at android.widget.FrameLayout.layoutChildren(FrameLayout.java:332)
    at android.widget.FrameLayout.onLayout(FrameLayout.java:270)
    at android.view.View.layout(View.java:23579)
    at android.view.ViewGroup.layout(ViewGroup.java:6412)
    at android.widget.LinearLayout.setChildFrame(LinearLayout.java:1829)
    at android.widget.LinearLayout.layoutVertical(LinearLayout.java:1673)
    at android.widget.LinearLayout.onLayout(LinearLayout.java:1582)
    at android.view.View.layout(View.java:23579)
    at android.view.ViewGroup.layout(ViewGroup.java:6412)
    at android.widget.FrameLayout.layoutChildren(FrameLayout.java:332)
    at android.widget.FrameLayout.onLayout(FrameLayout.java:270)
    at android.view.View.layout(View.java:23579)
    at android.view.ViewGroup.layout(ViewGroup.java:6412)
    at android.widget.LinearLayout.setChildFrame(LinearLayout.java:1829)
    at android.widget.LinearLayout.layoutVertical(LinearLayout.java:1673)
    at android.widget.LinearLayout.onLayout(LinearLayout.java:1582)
    at android.view.View.layout(View.java:23579)
    at android.view.ViewGroup.layout(ViewGroup.java:6412)
    at android.widget.FrameLayout.layoutChildren(FrameLayout.java:332)
    at android.widget.FrameLayout.onLayout(FrameLayout.java:270)
    at com.android.internal.policy.DecorView.onLayout(DecorView.java:798)
    at android.view.View.layout(View.java:23579)
    at android.view.ViewGroup.layout(ViewGroup.java:6412)
    at android.view.ViewRootImpl.performLayout(ViewRootImpl.java:3803)
    at android.view.ViewRootImpl.performTraversals(ViewRootImpl.java:3215)
    at android.view.ViewRootImpl.doTraversal(ViewRootImpl.java:2212)
    at android.view.ViewRootImpl$TraversalRunnable.run(ViewRootImpl.java:8973)
    at android.view.Choreographer$CallbackRecord.run(Choreographer.java:1211)
    at android.view.Choreographer$CallbackRecord.run(Choreographer.java:1219)
    at android.view.Choreographer.doCallbacks(Choreographer.java:892)
    at android.view.Choreographer.doFrame(Choreographer.java:825)
    at android.view.Choreographer$FrameDisplayEventReceiver.run(Choreographer.java:1194)
    at android.os.Handler.handleCallback(Handler.java:942)
    at android.os.Handler.dispatchMessage(Handler.java:99)
    at android.os.Looper.loopOnce(Looper.java:201)
    at android.os.Looper.loop(Looper.java:288)
    at android.app.ActivityThread.main(ActivityThread.java:7844)
    at java.lang.reflect.Method.invoke(Native Method)
    at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:548)
    at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:936)
 Caused by: java.lang.ExceptionInInitializerError
   at io.github.muntashirakon.AppManager.appops.AppOpsManager.permissionToOpCode(AppOpsManager.java:1843)
   at io.github.muntashirakon.AppManager.details.AppDetailsViewModel.loadAppOps(AppDetailsViewModel.java:1326)
   at io.github.muntashirakon.AppManager.details.AppDetailsViewModel.lambda$load$22$AppDetailsViewModel(AppDetailsViewModel.java:897)
   at io.github.muntashirakon.AppManager.details.-$$Lambda$AppDetailsViewModel$pWKoN_gf__X45S4rOvSG3-rZSro.run(Unknown Source:4)
   at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:463)
   at java.util.concurrent.FutureTask.run(FutureTask.java:264)
   at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1137)
   at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:637)
   at java.lang.Thread.run(Thread.java:1012)

Device Info:
App version: 3.0.0-alpha03-DEBUG
App version code: 402
Android build version: 8473704
Android release version: 12
Android SDK version: 32
Android build ID: TPB1.220310.029
Device brand: google
Device manufacturer: Google
Device name: oriole
Device model: Pixel 6
Device product name: oriole
Device hardware name: oriole
ABIs: [arm64-v8a, armeabi-v7a, armeabi]
ABIs (32bit): [armeabi-v7a, armeabi]
ABIs (64bit): [arm64-v8a]
System language: en-US
In-App Language: auto
Mode: auto
`
<details>

**Comment by MuntashirAkon** [comment]
> Ok nvm it crashes when I try to look at the app permissions and ops

This is due to new app ops added in A13. Fixing this will take some time as I am migrating to a new scheme that will fetch app ops dynamically preventing crashes each time new app ops are added.

**Comment by MuntashirAkon** [comment]
Could anybody send me framework.jar from the latest public beta? Thanks in advance.

**Comment by nate-moo** [comment]
I'm not on the latest but I'm on TPB3.220513.017 (which is the first revision of B3 I guess) if that's fine (https://files.catbox.moe/0wglms.7z), when I get home tomorrow I will be able to update and provide the newer framework.jar file

**Comment by MuntashirAkon** [comment]
It looks like the newer betas have the correct operating system and SDK info. So, we can now remove the workaround.

**Comment by nate-moo** [comment]
Here is TPB3.220513.017.B1 framework.jar:
https://files.catbox.moe/0wglms.7z

**Comment by nate-moo** [comment]
Here is TPB3.220610.004's framework.jar
https://drive.google.com/file/d/15Y6GAThx6zud1QWLFMLf-ANFbxCNy-Wx/view?usp=drivesdk

**Comment by MuntashirAkon** [comment]
> Ok nvm it crashes when I try to look at the app permissions and ops

Should be fixed in cbe5ed12bd96a3c54b819be484fef96a90e87e9b

I will also keep the workaround for SDK 33 detection until A13 is publicly available (cbe5ed12bd96a3c54b819be484fef96a90e87e9b).
